File: behaviorial/command/python/concrete_command.py


from interface import Command


class ToggleCommand(Command):
    def __init__(self, device):
        self.device = device

    def execute(self):
        if self.device.is_on():
            self.device.off()
        else:
            self.device.on()

    def undo(self):
        self.execute()

File: behaviorial/command/typescript/concrete-command.ts
import { Command } from './interface';
import { Device } from './receiver';

export class ToggleCommand implements Command {
	private device: Device;

	constructor(device: Device) {
		this.device = device;
	}

	execute(): void {
		if (this.device.isOn()) {
			this.device.off();
		} else {
			this.device.on();
		}
	}

	undo(): void {
		this.execute(); // Toggle back
	}
}
